### PR TITLE
bpo-45172: netbsd socketmodule CAN protocol constants addition.

### PR DESCRIPTION
**Pull request by devnexen**
https://bugs.python.org/issue45172

### CODE REVIEW

File: Modules/socketmodule.c [hunk 1]
```diff
@@ -7705,9 +7705,11 @@ PyInit__socket(void)
     PyModule_AddIntMacro(m, CAN_J1939);
 #endif
 #endif
-#ifdef HAVE_LINUX_CAN_RAW_H
+#if defined(HAVE_LINUX_CAN_RAW_H) || defined(HAVE_NETCAN_CAN_H)
     PyModule_AddIntMacro(m, CAN_RAW_FILTER);
+#if defined(HAVE_LINUX_CAN_RAW_H)
     PyModule_AddIntMacro(m, CAN_RAW_ERR_FILTER);
+#endif
```
Comment (vstinner) [inline Modules/socketmodule.c]: Maybe move CAN_RAW_ERR_FILTER above in a separated block to avoid nested #if.

File: configure.ac [hunk 2]
```diff
@@ -2314,6 +2314,13 @@ AC_CHECK_HEADERS(linux/can.h linux/can/bcm.h linux/can/j1939.h linux/can/raw.h,,
 #endif
 ])
 
+# CAN protocol constants on NetBSD are defined in this header
+AC_CHECK_HEADERS(netcan/can.h,,,[
+#ifdef HAVE_SYS_SOCKET_H
+#include <sys/socket.h>
+#endif
```
Comment (erlend-aasland) [inline configure.ac]: Do you really need to include `<sys/socket.h>` for the header check?

**Review by erlend-aasland — COMMENTED** [review]
@devnexen, can you resolve conflicts and address Victor's review?

File: configure.ac (hunk 2 above)
Comment (devnexen) [reply configure.ac]: not anymore, check done earlier already anyway.

File: pyconfig.h.in [hunk 3]
```diff
@@ -688,6 +688,9 @@
 /* Define if compiling using Linux 4.1 or later. */
 #undef HAVE_LINUX_CAN_RAW_JOIN_FILTERS
 
+/* Define to 1 if you have the <netcan/can.h> header file. */
+#undef HAVE_NETCAN_CAN_H
+
```
Comment (erlend-aasland) [inline pyconfig.h.in]:
```suggestion
```

**Comment by erlend-aasland** [comment]
Please avoid force-pushing PRs. I'm now unable to see the changes between your first version of the PR and the new version. It makes the reviewers job easier if you use `git fetch --all && git merge --no-ff main`.

**Review by erlend-aasland — APPROVED** [review]
Looks good to me. If Victor gives a thumbs up, I'll merge in a day or two.

File: Misc/NEWS.d/next/Library/2021-09-11-13-45-47.bpo-45172.T0LUgT.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Adds CAN protocol constants for NetBSD to :mod:`socket`.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Library/2021-09-11-13-45-47.bpo-45172.T0LUgT.rst]: The NEWS entry might be more specific: specify CAN_RAW_ERR_FILTER.

**Review by vstinner — APPROVED** [review]
LGTM. @erlend-aasland: you can merge it if you're fine with the current NEWS entry.

File: Misc/NEWS.d/next/Library/2021-09-11-13-45-47.bpo-45172.T0LUgT.rst (hunk 4 above)
Comment (erlend-aasland) [reply Misc/NEWS.d/next/Library/2021-09-11-13-45-47.bpo-45172.T0LUgT.rst]: I agree with Victor; explicit is better than implicit.

**Comment by erlend-aasland** [comment]
> LGTM. @erlend-aasland: you can merge it if you're fine with the current NEWS entry.

Thanks, I'll merge after the NEWS entry has been amended.

**Comment by devnexen** [comment]
I updated the entry accordingly.

**Comment by devnexen** [comment]
Ah I did not notice that one, well if this is supported already then my PR is pointless now.

**Comment by erlend-aasland** [comment]
Nonetheless, thanks for you interest in improving CPython :)

**Comment by vstinner** [comment]
Well, the important thing is that Python got CAN support on NetBSD ;-) https://github.com/python/cpython/commit/40fcd16889028bd3cd2289e0f8a2af43f17a5824 is part of Python 3.11: you can test the beta1!

Yeah, thanks for your interest in enhancing Python!

**Pull request closed without merging**